cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreEndpointContextInfo.java

```diff
@@ -0,0 +1,193 @@
+     *
+     * @param oscoreCtx the OSCORE context used for this message
+     * @param request the request to set the endpoint context for
+     */
+    public static void receivingRequest(OSCoreCtx oscoreCtx, Request request) {
+        setInfoIncoming(oscoreCtx, request);
+    }
+
+    /**
+     * Sets information in a destination endpoint context for outgoing responses.
+     *
+     * @param oscoreCtx the OSCORE context used for this message
+     * @param response the response to set the endpoint context for
+     */
+    public static void sendingResponse(OSCoreCtx oscoreCtx, Response response) {
+        setInfoOutgoing(oscoreCtx, response);
+    }
+
+    /**
+     * Sets information in a destination endpoint context for incoming responses.
+     *
+     * @param oscoreCtx the OSCORE context used for this message
+     * @param response the response to set the endpoint context for
+     */
+    public static void receivingResponse(OSCoreCtx oscoreCtx, Response response) {
+        setInfoIncoming(oscoreCtx, response);
+    }
+
+    /**
+     * Adds strings with information about the OSCORE context used to this
+     * destination endpoint context for outgoing messages.
+     *
+     * @param oscoreCtx the OSCORE context used for this message
+     * @param message the message to set information in the destination endpoint context for
+     */
+    private static void setInfoOutgoing(OSCoreCtx oscoreCtx, Message message) {
+
+        //Create new MapBasedEndpointContext for destination endpoint context with string values for OSCORE added
+        EndpointContext newEndpointContext = setInfo(oscoreCtx, message.getDestinationContext());
+        message.setDestinationContext(newEndpointContext);
+    }
+
+    /**
+     * Adds strings with information about the OSCORE context used to this
+     * source endpoint context for incoming messages.
+     *
+     * @param oscoreCtx the OSCORE context used for this message
+     * @param message the message to set information in the source endpoint context for
+     */
+    private static void setInfoIncoming(OSCoreCtx oscoreCtx, Message message) {
+
+        //Create new MapBasedEndpointContext for source endpoint context with string values for OSCORE added
+        EndpointContext newEndpointContext = setInfo(oscoreCtx, message.getSourceContext());
+        message.setSourceContext(newEndpointContext);
+    }
+
+    /**
+     * Adds strings with information about the OSCORE context used to this
+     * endpoint context (creating a new one).
+     *
+     * @param oscoreCtx the OSCORE context used
+     * @param endpointContext the original endpoint context to set information based on
+     *
+     * @return the new endpoint context
+     */
+    private static MapBasedEndpointContext setInfo(OSCoreCtx oscoreCtx, EndpointContext endpointContext) {
+
+    	//If endpoint context is not set, keep it unset
+    	if (endpointContext == null) {
```

Comment:
I'm OK with that check here in `setInfo`.  

Just to mention:
That context is set during sending in `CoapEndpoint`:

```
@Override
public void sendRequest(final Request request) {
	if (!started) {
		request.cancel();
		return;
	}
	// create context, if not already set
	request.prepareDestinationContext();
```

and the connector provides it in the RawData for received Messages.

If the unit tests creates there own Messages and want to use the contexts, they must ensure, that the contexts are set before processing the Messages in the CoapStack..

Reply:
Okay, thanks for the information. I will update the JUnit tests to set an endpoint context properly, since it seems some of them do not do this currently. They are calling methods from the ObjectSecurityLayer like prepareReceive() which will end up calling the method for adding information to the endpoint context, but none has been set which would cause the null pointer exception. If I update the JUnit tests to solve this is it alright to make that into a new PR?

Reply:
FMPOV, you have been asked to change this PR so often, that I think we should go with it now. If you want to spend your time into adjust the unit tests, you may do this in a future PR.

Reply:
> FMPOV, you have been asked to change this PR so often, that I think we should go with it now. If you want to spend your time into adjust the unit tests, you may do this in a future PR.

Yes, if I work on updating the JUnit tests to make sure they set endpoint contexts it will be in a new PR.